### PR TITLE
Introduce Error Prone

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR adds Error Prone to the project and replaces `@Nullable` from `javax.annotations` with the one from the Checker Framework.

It also starts eliminating the `final` qualifier for the local variables, since under Java 8 they are effectively final. We also want to improve the readability in the light of introduction of the `var` declaration in Java 10. So, this request also introduced an improved layout for calling builder methods.
For information about the `var` and `final` matters, please see [this page](http://mail.openjdk.java.net/pipermail/platform-jep-discuss/2016-December/000066.html).

Spine version advances to `0.10.41-SNAPSHOT`.

### CODE REVIEW

**Comment by alexander-yevsyukov** [comment]
@dmdashenkov, PTAL.

@armiol, FYI.

File: client/src/main/java/io/spine/client/ActorRequestFactory.java [hunk 1]
```diff
@@ -137,11 +135,13 @@ protected CommandContext createCommandContext() {
      * <p>Sets the timestamp value to the
      * {@linkplain io.spine.base.Time#getCurrentTime() current time}.
      */
+    @SuppressWarnings("CheckReturnValue") // calling builder
     ActorContext actorContext() {
-        final ActorContext.Builder builder = ActorContext.newBuilder()
-                                                         .setActor(actor)
-                                                         .setTimestamp(getCurrentTime())
-                                                         .setZoneOffset(zoneOffset);
+        ActorContext.Builder builder = ActorContext
```
Comment (alexander-yevsyukov) [inline client/src/main/java/io/spine/client/ActorRequestFactory.java]: @dmdashenkov, it's the new layout for calling builders. In Java 11, we'd  have `var` instead of type.

File: server/src/main/java/io/spine/server/commandbus/CommandScheduler.java [hunk 2]
```diff
@@ -57,11 +56,9 @@
 
     private boolean isActive = true;
 
-    @Nullable
-    private CommandBus commandBus;
+    private @Nullable CommandBus commandBus;
 
-    @Nullable
-    private Rescheduler rescheduler;
+    private @Nullable Rescheduler rescheduler;
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/commandbus/CommandScheduler.java]: Why do we introduce this change at all?

What's wrong with the old approach?

File: server/src/main/java/io/spine/server/entity/RecordBasedRepository.java [hunk 3]
```diff
@@ -465,7 +457,7 @@ public I apply(@Nullable EntityId input) {
             final Message idAsMessage = unpack(idAsAny);
 
             @SuppressWarnings("unchecked")
-                // As the message class is the same as expected, the conversion is safe.
+            // As the message class is the same as expected, the conversion is safe.
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/RecordBasedRepository.java]: Is this change intended?

File: server/src/main/java/io/spine/server/entity/Repository.java [hunk 4]
```diff
@@ -104,7 +101,7 @@ protected Repository() {
         if (entityClass == null) {
             @SuppressWarnings("unchecked") // The type is ensured by the declaration of this class.
             final Class<E> cast =
-                    (Class<E>)ENTITY.getArgumentIn((Class<? extends Repository<I, E>>) getClass());
+                    (Class<E>)ENTITY.getArgumentIn(getClass());
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/Repository.java]: Please add a space after `(Class<E>)`.

File: .idea/codeStyles/Project.xml [hunk 5]
```diff
@@ -54,4 +57,4 @@
       </arrangement>
     </codeStyleSettings>
   </code_scheme>
-</component>
+</component>
```
Comment (dmdashenkov) [inline .idea/codeStyles/Project.xml]: Please keep the EOFs.

File: server/src/test/java/io/spine/server/projection/ProjectionTransactionShould.java [hunk 6]
```diff
@@ -52,10 +52,10 @@
...
 public class ProjectionTransactionShould
         extends TransactionShould<ProjectId,
-                                  Projection<ProjectId, Project,
-                                          ProjectionTransactionShould.PatchedProjectBuilder>,
-                                  Project,
-                                  ProjectionTransactionShould.PatchedProjectBuilder> {
+        Projection<ProjectId, Project,
+                ProjectionTransactionShould.PatchedProjectBuilder>,
+        Project,
+        ProjectionTransactionShould.PatchedProjectBuilder> {
```
Comment (dmdashenkov) [inline server/src/test/java/io/spine/server/projection/ProjectionTransactionShould.java]: Please revert this change.

File: server/src/test/java/io/spine/server/stand/StandPostShould.java [hunk 7]
```diff
@@ -163,103 +107,149 @@ private static void checkUpdatesDelivery(boolean isConcurrent,
...
-                    // Handle null event dispatching after the command is handled.
-
-                    // Check if this error is caused by returning nuu or empty list after
-                    // command processing.
-                    // Proceed crash if it's not.
-                    if (!e.getMessage()
-                          .contains("No record found for command ID: EMPTY")) {
-                        throw e;
-                    }
+        return context -> {
```
Comment (dmdashenkov) [inline server/src/test/java/io/spine/server/stand/StandPostShould.java]: It would be better to inline this lambda body with the method body and use the method reference.

File: server/src/test/java/io/spine/server/stand/StandPostShould.java [hunk 8]
```diff
@@ -163,103 +107,149 @@ private static void checkUpdatesDelivery(boolean isConcurrent,
...
+                                                      .equals(entityId);
+                    boolean versionMatches = version.equals(argument.getEntityVersion()
+                                                                    .orNull());
+                    boolean stateMatches = argument.getMessage()
+                                                   .equals(state);
+                    return entityIdMatches &&
+                            versionMatches &&
+                            stateMatches;
+                };
+        verify(stand).update(ArgumentMatchers.argThat(argumentMatcher));
```
Comment (dmdashenkov) [inline server/src/test/java/io/spine/server/stand/StandPostShould.java]: Please add the static import for `argThat`.

File: server/src/test/java/io/spine/server/storage/RecordStorageShould.java [hunk 9]
```diff
@@ -284,23 +275,23 @@ public void write_none_storage_fields_is_none_passed() {
...
         storage.write(initial);
 
-        final Collection<EntityRecord> actual = newArrayList(
+        Collection<EntityRecord> actual = newArrayList(
                 storage.readMultiple(initial.keySet())
         );
-        final Collection<EntityRecord> expected = Collections2.transform(initial.values(),
-                                                                         RECORD_EXTRACTOR_FUNCTION);
+        Collection<EntityRecord> expected =
+                Collections2.transform(initial.values(), EntityRecordWithColumns::getRecord);
```
Comment (dmdashenkov) [inline server/src/test/java/io/spine/server/storage/RecordStorageShould.java]: Please replace with Stream API.

File: server/src/test/java/io/spine/server/storage/RecordStorageShould.java [hunk 10]
```diff
@@ -890,6 +894,7 @@ private void delete() {
         UNRECOGNIZED
     }
 
+
```
Comment (dmdashenkov) [inline server/src/test/java/io/spine/server/storage/RecordStorageShould.java]: This line is redundant.

File: testutil-server/src/main/java/io/spine/server/entity/given/Given.java [hunk 11]
```diff
@@ -62,49 +62,57 @@ private Given() {
...
      */
-    public static <A extends AggregatePart<I, S, ?, R>,
-                   I,
-                   S extends Message,
-                   R extends AggregateRoot<I>>
+    public static <
+            A extends AggregatePart<I, S, ?, R>,
+            I,
+            S extends Message,
+            R extends AggregateRoot<I>
```
Comment (dmdashenkov) [inline testutil-server/src/main/java/io/spine/server/entity/given/Given.java]: Please add the padding.

File: testutil-server/src/test/java/io/spine/server/command/CommandTestShould.java [hunk 12]
```diff
@@ -53,14 +59,34 @@
...
     }
 
+    /**
+     * Creates a test instance of {@code ActorRequestFactory}.
+     *
+     * <p>The factory gets:
+     * <ul>
+     * <li>generated {@code UserId} for the actor.
+     * <li>UTC zone offset
+     * <li>{@code TenantId} based on the simple name of the passed class.
```
Comment (dmdashenkov) [inline testutil-server/src/test/java/io/spine/server/command/CommandTestShould.java]: Please inline the tags properly.

File: server/src/main/java/io/spine/server/commandbus/CommandScheduler.java (hunk 2 above)
Comment (alexander-yevsyukov) [reply server/src/main/java/io/spine/server/commandbus/CommandScheduler.java]: See [this issue](https://github.com/SpineEventEngine/base/issues/108) on this.

File: server/src/main/java/io/spine/server/commandstore/CEntity.java [hunk 13]
```diff
@@ -65,51 +65,58 @@ static CEntity createForError(Command command, Error error) {
...
+        CommandRecord.Builder builder = stateBuilder();
         builder.getStatusBuilder()
                .setCode(ERROR)
                .setError(error);
-        final CommandRecord record = builder.build();
+        CommandRecord record = builder.build();
         updateState(record);
     }
 
+    @SuppressWarnings("CheckReturnValue") // calling builder
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply server/src/main/java/io/spine/server/commandstore/CEntity.java]: There's in [an issue](https://github.com/SpineEventEngine/base/issues/117) on this already. Would you pitch in?

File: server/src/main/java/io/spine/server/entity/EventPlayingEntity.java [hunk 14]
```diff
@@ -58,8 +57,8 @@
      */
     private volatile boolean stateChanged;
 
-    @Nullable
-    private volatile Transaction<I, ? extends EventPlayingEntity<I, S, B>, S, B> transaction;
+    private volatile
+    @Nullable Transaction<I, ? extends EventPlayingEntity<I, S, B>, S, B> transaction;
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply server/src/main/java/io/spine/server/entity/EventPlayingEntity.java]: And it's not.

**Comment by alexander-yevsyukov** [comment]
@dmdashenkov, we discussed the matter of `final` vocally already. For more information about `final` and `var` matter, please see [this page](http://mail.openjdk.java.net/pipermail/platform-jep-discuss/2016-December/000066.html).

File: .idea/codeStyles/Project.xml (hunk 5 above)
Comment (alexander-yevsyukov) [reply .idea/codeStyles/Project.xml]: This file is IDEA-generated. I'll change it now, but with the new code style change, it's going to be overwritten again.

File: server/src/main/java/io/spine/server/projection/ProjectionRepository.java [hunk 15]
```diff
@@ -219,37 +233,36 @@ protected final Stand getStand() {
      */
     protected ProjectionStorage<I> projectionStorage() {
         @SuppressWarnings("unchecked") // OK as we control the creation in createStorage().
-        final ProjectionStorage<I> storage = (ProjectionStorage<I>) getStorage();
+                ProjectionStorage<I> storage = (ProjectionStorage<I>) getStorage();
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply server/src/main/java/io/spine/server/projection/ProjectionRepository.java]: It's the result of IDEA formatting. I have not seen an option that could prevent that.
I think we should adopt a pre-checkin automated formatting by a tool like [Google Java Format](https://github.com/google/google-java-format). But GJF does not allow customization, and we do have our style modifications to the Google Java style.

File: server/src/main/java/io/spine/server/storage/memory/InMemoryStandStorage.java [hunk 16]
```diff
@@ -91,13 +90,10 @@ public static Builder newBuilder() {
                                                        final TypeUrl expectedType) {
         final Iterator<EntityRecord> result =
                 filter(records,
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply server/src/main/java/io/spine/server/storage/memory/InMemoryStandStorage.java]: I'd keep it the way it's now since the method returns `Iterator`.

File: server/src/test/java/io/spine/server/outbus/enrich/given/EnricherBuilderTestEnv.java [hunk 17]
```diff
@@ -51,24 +49,24 @@ private Enrichment() {}
...
-                   .add(ProjectId.class, UserId.class, new GetProjectOwnerId())
-                   .add(EventId.class, String.class, EVENT_ID_TO_STRING)
-                   .add(Timestamp.class, String.class, TIMESTAMP_TO_STRING)
-                   .add(CommandContext.class, String.class, CMD_CONTEXT_TO_STRING)
-                   .add(Any.class, String.class, ANY_TO_STRING)
-                   .add(Integer.class, String.class, VERSION_TO_STRING)
-                   .add(String.class, ZoneOffset.class, STRING_TO_ZONE_OFFSET)
-                   .add(String.class, PersonName.class, STRING_TO_PERSON_NAME)
-                   .add(String.class, Integer.class, STRING_TO_INT);
+            EventEnricher.Builder builder = EventEnricher
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply server/src/test/java/io/spine/server/outbus/enrich/given/EnricherBuilderTestEnv.java]: We'd do that in a separate PR. Please file an issue for that.

**Comment by alexander-yevsyukov** [comment]
@dmdashenkov, PTAL.

File: server/src/test/java/io/spine/server/outbus/enrich/given/EnricherBuilderTestEnv.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (dmdashenkov) [reply server/src/test/java/io/spine/server/outbus/enrich/given/EnricherBuilderTestEnv.java]: #717

File: server/src/main/java/io/spine/server/entity/storage/Methods.java [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+
+    static String nameFromGetter(Method getter) {
+        Matcher prefixMatcher = GETTER_PREFIX_PATTERN.matcher(getter.getName());
+        String nameWithoutPrefix = prefixMatcher.replaceFirst("");
+        String result =
+                Character.toLowerCase(nameWithoutPrefix.charAt(0)) + nameWithoutPrefix.substring(1);
+        return result;
+    }
+
+    static boolean hasNullableReturn(Method getter) {
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Methods.java]: Maybe, `mayReturnNull(Method)`?

File: server/src/main/java/io/spine/server/entity/storage/Methods.java [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    static void checkGetter(Method getter) {
+        checkNotNull(getter);
+        checkArgument(GETTER_PREFIX_PATTERN.matcher(getter.getName())
+                                           .find()
+                              && getter.getParameterTypes().length == 0,
+                      "Method `%s` is not a getter.", getter);
+        checkArgument(getAnnotatedVersion(getter).isPresent(),
+                      format("Entity column getter should be annotated with `%s`.",
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Methods.java]: Please remove redundant `format`, as `checkArgument` is capable of string formatting on its own.

File: server/src/main/java/io/spine/server/entity/storage/Methods.java [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+                             Column.class.getName()));
+        int modifiers = getter.getModifiers();
+        checkArgument(isPublic(modifiers) && !isStatic(modifiers),
+                      "Entity column getter should be public instance method.");
+        Class<?> returnType = getter.getReturnType();
+        Class<?> wrapped = wrap(returnType);
+        checkArgument(Serializable.class.isAssignableFrom(wrapped),
+                      format("Cannot create column of non-serializable type %s by method %s.",
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Methods.java]: Same as above.

File: server/src/main/java/io/spine/server/event/EventStore.java [hunk 21]
```diff
@@ -84,10 +84,9 @@ public static ServiceBuilder newServiceBuilder() {
 
     private static void ensureSameTenant(Iterable<Event> events) {
         checkNotNull(events);
-        Set<TenantId> tenants =
-                StreamSupport.stream(events.spliterator(), false)
-                             .map(Events::getTenantId)
-                             .collect(Collectors.toSet());
+        Set<TenantId> tenants = Streams.stream(events)
+                                       .map(Events::getTenantId)
+                                       .collect(Collectors.toSet());
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/event/EventStore.java]: Please add a static import for `toSet()`.

File: server/src/test/java/io/spine/server/aggregate/AggregateStorageShould.java [hunk 22]
```diff
@@ -110,12 +104,14 @@ public void setUpAbstractStorageTest() {
...
         List<AggregateEventRecord> records = sequenceFor(id);
-        List<Event> expectedEvents = transform(records, TO_EVENT);
-        AggregateStateRecord aggregateStateRecord =
-                AggregateStateRecord.newBuilder()
-                                    .addAllEvent(expectedEvents)
-                                    .build();
-        return aggregateStateRecord;
+        List<Event> expectedEvents = records.stream()
+                                            .map(TO_EVENT)
+                                            .collect(Collectors.toList());
```
Comment (dmdashenkov) [inline server/src/test/java/io/spine/server/aggregate/AggregateStorageShould.java]: Please add a static import for `toList()`.

File: server/src/test/java/io/spine/server/entity/TransactionShould.java [hunk 23]
```diff
@@ -69,7 +69,33 @@
...
+                .build();
+        return newArrayList(expectedViolation);
+    }
+
+    private static boolean checkPhase(Event event, Phase phase, boolean successful) {
+        Message eventMessage = Events.getMessage(event);
+        return eventMessage.equals(phase.getEventMessage())
+                && event.getContext()
+                        .equals(phase.getContext())
+                && phase.isSuccessful() == successful;
```
Comment (dmdashenkov) [inline server/src/test/java/io/spine/server/entity/TransactionShould.java]: Please split the expression into 3 with 3 local variables to make it more debuggable.

File: server/src/main/java/io/spine/server/aggregate/ReadOperation.java [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+     *
+     * @param record the record to check
+     * @throws IllegalStateException if the {@link AggregateStateRecord} is not valid
+     */
+    private static void checkRecord(AggregateStateRecord record) throws IllegalStateException {
+        boolean snapshotIsNotSet = !record.hasSnapshot();
+        boolean noEvents = record.getEventList()
+                                 .isEmpty();
+        if (noEvents && snapshotIsNotSet) {
+            throw new IllegalStateException("AggregateStateRecord instance should have either "
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply server/src/main/java/io/spine/server/aggregate/ReadOperation.java]: No need here as we're not passing arguments to a formatted string.

File: server/src/main/java/io/spine/server/commandbus/CommandBus.java [hunk 25]
```diff
@@ -353,8 +353,8 @@ protected CommandDispatcherRegistry registry() {
          * the current runtime environment.
          */
         private static boolean detectThreadsAllowed() {
-            final boolean appEngine = ServerEnvironment.getInstance()
-                                                       .isAppEngine();
+            boolean appEngine = ServerEnvironment.getInstance()
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply server/src/main/java/io/spine/server/commandbus/CommandBus.java]: This matter needs more attention across the whole codebase. Please file an issue for this and link it into this PR.

**Comment by alexander-yevsyukov** [comment]
@dmdashenkov, PTAL.

File: client/src/test/java/io/spine/core/CommandsTest.java [hunk 26]
```diff
@@ -192,49 +193,47 @@ void createWereBetweenPredicate() {
...
+                                                 twoMinsAgo,
+                                                 thirtySecondsAgo,
+                                                 twentySecondsAgo,
+                                                 fiveSecondsAgo)
+                                             .filter(wereWithinPeriod(minutesAgo(3), secondsAgo(10))
+                                                             ::apply)
+                                             .collect(toList());
 
             assertEquals(3, StreamSupport.stream(filter.spliterator(), false)
                                          .count());
```
Comment (dmdashenkov) [inline client/src/test/java/io/spine/core/CommandsTest.java]: This is still strange.

First, we generate a stream, then convert it to list and then back to stream, and, at last, obtain `count()`.

File: client/src/test/java/io/spine/core/CommandsTest.java [hunk 27]
```diff
@@ -192,49 +193,47 @@ void createWereBetweenPredicate() {
...
-        @Test
-        @DisplayName("consider command not scheduled when no scheduling options are present")
-        void recognizeNotScheduled() {
-            Command cmd = requestFactory.createCommand(StringValue.getDefaultInstance());
-            assertFalse(Commands.isScheduled(cmd));
-        }
+    @Test
+    @DisplayName("consider command not scheduled when no scheduling options are present")
+    void recognizeNotScheduled() {
+        final Command cmd = requestFactory.createCommand(StringValue.getDefaultInstance());
```
Comment (dmdashenkov) [inline client/src/test/java/io/spine/core/CommandsTest.java]: Why do we use `final` here again?
